### PR TITLE
Make create order api and write unit test.

### PR DESCRIPTION
**Pull request by virsavik**
- Make create order API.
- Write unit test for functions at handler, service and repository.

### CODE REVIEW

File: internal/handler/rest/v1/order.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+func validateOrderRequest(r OrderRequest) (order.OrderInput, error) {
+	if r.UserID <= 0 {
+		return order.OrderInput{}, ErrInvalidUserID
+	}
+	if len(r.Items) == 0 {
+		return order.OrderInput{}, ErrItemsIsRequired
+	}
+	items := make([]order.OrderItemInput, len(r.Items))
+	for i, item := range r.Items {
+		if item.ProductID == 0 {
```
Comment (virsavik) [inline internal/handler/rest/v1/order.go:33]: remove it

File: internal/handler/rest/v1/order.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	switch err.Error() {
+	case order.ErrUserIsNotExist.Error():
+		utils.WriteJSONResponse(w, http.StatusBadRequest, ErrUserIsNotExist)
+	case order.ErrProductIsNotExist.Error():
+		utils.WriteJSONResponse(w, http.StatusBadRequest, ErrProductIsNotExist)
+	default:
+		utils.WriteJSONResponse(w, http.StatusInternalServerError, ErrInternalServerError)
+	}
+}
+
```
Comment (virsavik) [inline internal/handler/rest/v1/order.go:69]: Write desc

File: internal/handler/rest/v1/order.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+		return
+	}
+
+	// Create order
+	if err = h.orderServ.CreateOrder(r.Context(), input); err != nil {
+		handleOrderError(w, err)
+		return
+	}
+
+	utils.WriteJSONResponse(w, http.StatusCreated, "Create order successfully")
```
Comment (virsavik) [inline internal/handler/rest/v1/order.go:91]: Created order...

File: internal/service/order/order.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		}
+		input.Items[i].ProductName = product.Title
+		input.Items[i].ProductPrice = product.Price
+	}
+
+	err = o.repo.Tx(ctx, func(tx *sql.Tx) error {
+		// Generate order number
+		orderNumber, err := uuid.NewV4()
+		if err != nil {
+			return err
```
Comment (virsavik) [inline internal/service/order/order.go:63]: Detail error

File: internal/service/order/order.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		// Create order
+		order, err := o.repo.Order().CreateOrder(ctx, tx, model.Order{
+			OrderNumber: orderNumber.String(),
+			OrderDate:   time.Now(),
+			Status:      string(OrderStatusNew),
+			Note:        input.Note,
+			UserID:      input.UserID,
+		})
+		if err != nil {
+			return err
```
Comment (virsavik) [inline internal/service/order/order.go:75]: fmt errorf

File: cmd/serverd/router/router.go [hunk 6]
```diff
@@ -21,7 +22,8 @@ func InitRouter(db *sql.DB) *chi.Mux {
 	repo := repository.New(db)
 	userServ := user.New(repo)
 	productServ := productService.New(repo)
-	h := v1.NewHandler(userServ, productServ)
+	orderServ := orderService.New(repo)
```
Comment (vinhnv1) [inline cmd/serverd/router/router.go:25]:
```suggestion
	orderServ := order.New(repo)
```

File: go.mod [hunk 7]
```diff
@@ -6,7 +6,11 @@ require (
 	github.com/99designs/gqlgen v0.17.13
 	github.com/friendsofgo/errors v0.9.2
 	github.com/go-chi/chi/v5 v5.0.7
+<<<<<<< HEAD
 	github.com/jordan-wright/email v4.0.1-0.20210109023952-943e75fe5223+incompatible
+=======
+	github.com/gofrs/uuid v3.2.0+incompatible
+>>>>>>> 7a10d15 (solve convention problem.)
```
Comment (vinhnv1) [inline go.mod:13]: Pls resolve this confilct

File: internal/handler/rest/v1/order.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	return order.OrderInput{
+		Note:   r.Note,
+		UserID: r.UserID,
+		Items:  items,
+	}, nil
+}
+
+// handleOrderError handles error from order service
+func handleOrderError(w http.ResponseWriter, err error) {
+	switch err.Error() {
```
Comment (vinhnv1) [inline internal/handler/rest/v1/order.go:58]:
```suggestion
	switch err {
```

File: internal/service/order/order.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+		return err
+	}
+	if !existed {
+		return ErrUserIsNotExist
+	}
+
+	// Check exists product by order item product_id
+	for i, item := range input.Items {
+		product, err := o.repo.Product().GetProduct(ctx, item.ProductID)
+		if err == sql.ErrNoRows {
```
Comment (vinhnv1) [inline internal/service/order/order.go:51]: if errors.Is(err, sql.ErrNoRows) {..}

File: internal/service/order/order.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+		input.Items[i].ProductName = product.Title
+		input.Items[i].ProductPrice = product.Price
+	}
+
+	if err = o.repo.Tx(ctx, func(tx *sql.Tx) error {
+		// Generate order number
+		orderNumber, err := uuid.NewV4()
+		if err != nil {
+			fmt.Printf("Error when generate order number: %v\n", err)
+			return err
```
Comment (vinhnv1) [inline internal/service/order/order.go:65]:
```suggestion
			return fmt.Errorf("Error when generate order number: %v\n", err)
```

File: cmd/serverd/router/router.go (hunk 6 above)
Comment (virsavik) [reply cmd/serverd/router/router.go:25]: done

**Pull request merged**